File: libs/closure_agent/src/spec.cpp
// spec.cpp — ActionSpec behaviour, byte-identical problem strings and JSON
// projections against the frozen Python oracle (agent_oracle.json).
#include <pwb/closure_agent/spec.hpp>

#include <pwb/providers/contracts.hpp>
#include <pwb/providers/schema.hpp>

namespace pwb::closure_agent {

std::string to_string(ActionRisk risk) {
    switch (risk) {
        case ActionRisk::Read: return "read";
        case ActionRisk::Compute: return "compute";
        case ActionRisk::Write: return "write";
        case ActionRisk::Destructive: return "destructive";
    }
    return "read";
}

std::optional<ActionRisk> action_risk_from_string(const std::string& value) {
    if (value == "read") return ActionRisk::Read;
    if (value == "compute") return ActionRisk::Compute;
    if (value == "write") return ActionRisk::Write;
    if (value == "destructive") return ActionRisk::Destructive;
    return std::nullopt;
}

std::string to_string(ActionStatus status) {
    switch (status) {
        case ActionStatus::Success: return "success";
        case ActionStatus::Degraded: return "degraded";
        case ActionStatus::Failed: return "failed";
        case ActionStatus::Cancelled: return "cancelled";
        case ActionStatus::Rejected: return "rejected";
        case ActionStatus::Unavailable: return "unavailable";
    }
    return "failed";
}

bool is_positive_status(ActionStatus status) {
    return status == ActionStatus::Success || status == ActionStatus::Degraded;
}

bool is_known_typed_ref(const std::string& ref) {
    for (const auto& known : pwb::providers::typed_refs()) {
        if (known == ref) return true;
    }
    return false;
}

std::string ActionSpec::domain() const {
    const auto pos = action_id.find('.');
    return pos == std::string::npos ? action_id : action_id.substr(0, pos);
}

Json ActionSpec::effective_resource_profile() const {
    if (!resource_profile.is_null()) return resource_profile;
    Json profile = Json::object();
    profile["estimated_cpu_cores"] = 0.5;
    profile["estimated_ram_bytes"] = 0;
    profile["estimated_vram_bytes"] = 0;
    profile["io_weight"] = 0.5;
    return profile;
}

Json ActionSpec::to_dict() const {
    Json dict = Json::object();
    dict["action_id"] = action_id;
    dict["description"] = description;
    dict["input_schema"] = input_schema;
    dict["output_schema"] = output_schema;
    dict["risk"] = to_string(risk);
    dict["category"] = category;
    dict["resource_profile"] = effective_resource_profile();
    dict["required_context"] = required_context;
    dict["supports_cancel"] = supports_cancel;
    dict["provider_id"] = provider_id ? Json(*provider_id) : Json(nullptr);
    dict["side_effect_notes"] = side_effect_notes;
    dict["version"] = version;
    dict["deterministic"] = deterministic;
    dict["cacheable"] = cacheable;
    dict["idempotent"] = idempotent;
    dict["domain_tags"] = domain_tags;
    dict["input_refs"] = input_refs;
    dict["output_refs"] = output_refs;
    return dict;
}

Json ActionSpec::tool_schema() const {
    std::string derived = description + " [risk: " + to_string(risk);
    if (provider_id) derived += "; via provider " + *provider_id;
    derived += "]";
    // Python: action_id.replace(".", "__").
    std::string mapped;
    mapped.reserve(action_id.size() + 4);
    for (const char ch : action_id) {
        if (ch == '.') {
            mapped += "__";
        } else {
            mapped += ch;
        }
    }
    Json function = Json::object();
    function["name"] = mapped;
    function["description"] = derived;
    // Python: self.input_schema or {default object schema} (empty dict is
    // falsy).
    Json parameters =
        input_schema.is_object() && !input_schema.empty()
            ? input_schema
            : Json(Json::object());
    if (!input_schema.is_object() || input_schema.empty()) {
        parameters["type"] = "object";
        parameters["properties"] = Json::object();
        parameters["additionalProperties"] = false;
    }
    function["parameters"] = parameters;
    Json schema = Json::object();
    schema["type"] = "function";
    schema["function"] = function;
    return schema;
}

namespace {

// Python dict/get semantics: absent key -> Json(nullptr).
const Json& get(const Json& object, const char* key) {
    static const Json null_value = Json(nullptr);
    if (!object.is_object()) return null_value;
    const auto it = object.find(key);
    return it == object.end() ? null_value : *it;
}

bool is_known_type_name(const std::string& name) {
    return name == "object" || name == "array" || name == "string" ||
           name == "integer" || name == "number" || name == "boolean" ||
           name == "null";
}

}  // namespace

std::vector<std::string> validate_schema_shape(const Json& schema,
                                               const std::string& path) {
    std::vector<std::string> problems;
    if (!schema.is_object()) {
        problems.push_back(path + ": must be a dict (JSON schema)");
        return problems;
    }
    const Json& expected = get(schema, "type");
    if (!expected.is_null()) {
        std::vector<std::string> names;
        if (expected.is_array()) {
            for (const auto& entry : expected) names.push_back(entry.get<std::string>());
        } else {
            names.push_back(expected.get<std::string>());
        }
        for (const auto& name : names) {
            if (!is_known_type_name(name)) {
                problems.push_back(path + ": unknown type " +
                                   pwb::providers::python_repr(name));
            }
        }
    }
    const Json& enum_value = get(schema, "enum");
    if (!enum_value.is_null() && !enum_value.is_array()) {
        problems.push_back(path + ": enum must be a list");
    }
    for (const char* bound : {"minimum", "maximum"}) {
        const Json& value = get(schema, bound);
        if (!value.is_null() && !value.is_number()) {
            problems.push_back(path + ": " + bound + " must be numeric");
        }
    }
    for (const char* bound : {"minItems", "maxItems"}) {
        const Json& value = get(schema, bound);
        if (!value.is_null() && !value.is_number_integer()) {
            problems.push_back(path + ": " + bound + " must be an integer");
        }
    }
    const Json& properties = get(schema, "properties");
    if (!properties.is_null()) {
        if (!properties.is_object()) {
            problems.push_back(path + ".properties: must be a dict");
        } else {
            for (auto it = properties.begin(); it != properties.end(); ++it) {
                auto sub = validate_schema_shape(
                    it.value(), path + ".properties." + it.key());
                problems.insert(problems.end(), sub.begin(), sub.end());
            }
        }
    }
    const Json& required = get(schema, "required");
    if (!required.is_null()) {
        bool ok = required.is_array();
        if (ok) {
            for (const auto& entry : required) {
                if (!entry.is_string()) {
                    ok = false;
                    break;
                }
            }
        }
        if (!ok) problems.push_back(path + ".required: must be a list of strings");
    }
    const Json& items = get(schema, "items");
    if (!items.is_null()) {
        if (items.is_object()) {
            auto sub = validate_schema_shape(items, path + ".items");
            problems.insert(problems.end(), sub.begin(), sub.end());
        } else {
            problems.push_back(path + ".items: must be a dict");
        }
    }
    const Json& additional = get(schema, "additionalProperties");
    if (!additional.is_null() && !additional.is_boolean()) {
        problems.push_back(path + ".additionalProperties: must be a boolean");
    }
    return problems;
}

namespace {

// Python re.match(r"^[a-z][a-z0-9_]*\.[a-z][a-z0-9_]*$", id) — full match.
bool matches_action_id_pattern(const std::string& id) {
    const auto dot = id.find('.');
    if (dot == std::string::npos || id.find('.', dot + 1) != std::string::npos) {
        return false;
    }
    auto part_ok = [](const std::string& part) {
        if (part.empty()) return false;
        if (part[0] < 'a' || part[0] > 'z') return false;
        for (const char ch : part) {
            const bool ok = (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9') ||
                            ch == '_';
            if (!ok) return false;
        }
        return true;
    };
    return part_ok(id.substr(0, dot)) && part_ok(id.substr(dot + 1));
}

// Python re.match(r"^[0-9]+(\.[0-9]+){0,3}$", version) — full match:
// one to four dot-separated all-digit groups.
bool matches_version_pattern(const std::string& version) {
    std::size_t i = 0;
    auto digits = [&](std::size_t start) {
        std::size_t j = start;
        while (j < version.size() && version[j] >= '0' && version[j] <= '9') ++j;
        return j;
    };
    i = digits(0);
    if (i == 0) return false;
    int groups = 1;
    while (i < version.size()) {
        if (version[i] != '.') return false;
        const std::size_t next = digits(i + 1);
        if (next == i + 1) return false;
        i = next;
        ++groups;
    }
    return groups >= 1 && groups <= 4;
}

// Python re.match(r"^[a-z0-9][a-z0-9_.-]*$", tag) — full match.
bool matches_tag_pattern(const std::string& tag) {
    if (tag.empty()) return false;
    const auto head_ok = [](char ch) {
        return (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9');
    };
    if (!head_ok(tag[0])) return false;
    for (const char ch : tag) {
        const bool ok = (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9') ||
                        ch == '_' || ch == '.' || ch == '-';
        if (!ok) return false;
    }
    return true;
}

}  // namespace

std::vector<std::string> validate_action_spec(const ActionSpec& spec) {
    std::vector<std::string> problems;
    if (!matches_action_id_pattern(spec.action_id)) {
        problems.push_back("action_id " + pwb::providers::python_repr(spec.action_id) +
                           " must be '<domain>.<name>' lowercase dotted");
    }
    bool description_blank = true;
    for (const char ch : spec.description) {
        if (ch != ' ' && ch != '\t' && ch != '\n' && ch != '\r' && ch != '\v' &&
            ch != '\f') {
            description_blank = false;
            break;
        }
    }
    if (description_blank) {
        problems.push_back("description must be non-empty");
    }
    if (!spec.handler && !spec.provider_id) {
        problems.push_back("spec needs a handler or a provider_id to be executable");
    }
    if (!spec.input_schema.is_object()) {
        problems.push_back("input_schema must be a dict (JSON schema)");
    } else {
        const Json& top_type = get(spec.input_schema, "type");
        if (!top_type.is_null() &&
            (!top_type.is_string() || top_type.get<std::string>() != "object")) {
            problems.push_back("input_schema must describe an object at the top level");
        }
        auto shape = validate_schema_shape(spec.input_schema, "input_schema");
        problems.insert(problems.end(), shape.begin(), shape.end());
    }
    if (spec.output_schema.is_object() && !spec.output_schema.empty()) {
        auto shape = validate_schema_shape(spec.output_schema, "output_schema");
        problems.insert(problems.end(), shape.begin(), shape.end());
    }
    if (!spec.version.empty() && !matches_version_pattern(spec.version)) {
        problems.push_back("version " + pwb::providers::python_repr(spec.version) +
                           " must be numeric dotted (e.g. 1.0)");
    }
    if (spec.cacheable && !spec.deterministic) {
        problems.push_back(
            "cacheable requires deterministic (a cache hit replays the execution)");
    }
    if (spec.cacheable && spec.output_refs.empty()) {
        problems.push_back(
            "cacheable requires declared output_refs: a cache hit must "
            "re-materialise from catalog-resolvable outputs");
    }
    for (const auto& tag : spec.domain_tags) {
        if (!matches_tag_pattern(tag)) {
            problems.push_back("domain_tag " + pwb::providers::python_repr(tag) +
                               " must match ^[a-z0-9][a-z0-9_.-]*$");
        }
    }
    for (const auto& [role, refs] :
         {std::pair<const char*, const std::vector<std::string>&>{"input_refs",
                                                                 spec.input_refs},
          {"output_refs", spec.output_refs}}) {
        for (const auto& ref : refs) {
            if (!is_known_typed_ref(ref)) {
                std::string known = "[";
                const auto& typed = pwb::providers::typed_refs();
                for (std::size_t i = 0; i < typed.size(); ++i) {
                    if (i) known += ", ";
                    known += pwb::providers::python_repr(typed[i]);
                }
                known += "]";
                problems.push_back(std::string(role) + " entry " +
                                   pwb::providers::python_repr(ref) +
                                   " is not a known typed ref " + known);
            }
        }
    }
    // resource_profile: null -> the Python default profile (always valid);
    // an explicit profile is checked like the Python dict.get defaults.
    if (!spec.resource_profile.is_null()) {
        if (!spec.resource_profile.is_object()) {
            problems.push_back("resource_profile must be a dict");
        } else {
            const Json& cores = get(spec.resource_profile, "estimated_cpu_cores");
            const double cores_value =
                cores.is_number() ? cores.get<double>() : 0.5;
            if (cores_value <= 0.0) {
                problems.push_back(
                    "resource_profile.estimated_cpu_cores must be > 0");
            }
            const Json& io = get(spec.resource_profile, "io_weight");
            const double io_value = io.is_number() ? io.get<double>() : 0.5;
            if (io_value < 0.0) {
                problems.push_back("resource_profile.io_weight must be >= 0");
            }
            const Json& temp = get(spec.resource_profile, "estimated_temp_bytes");
            if (!temp.is_null() && !temp.is_number()) {
                problems.push_back(
                    "resource_profile.estimated_temp_bytes must be numeric");
            }
        }
    }
    return problems;
}

}  // namespace pwb::closure_agent

File: legacy/python_reference/product/paleo_workbench/harness/spec.py
"""Geological harness action specifications (P2-C, Harness 2.0 contracts).

An :class:`ActionSpec` is the single source of truth for one stable,
professional geological action. From it derive — never by hand, never a
second copy — the runtime validation, the machine-readable agent tool
schema, and the docs. Actions are coarse-grained and domain-semantic
(``map.create_factor_map``), not UI gestures (no findChild, no clicks).

Harness 2.0 additions (backward-compatible constructor arguments):

- ``version``: the action contract version — part of every cache key and
  receipt, bumped whenever behaviour or schema changes meaning;
- ``deterministic`` / ``cacheable`` / ``idempotent``: the reproducibility
  vocabulary. ``cacheable`` *requires* ``deterministic`` and a non-empty
  ``output_refs`` declaration — a cache hit must re-materialise from
  catalog-resolvable outputs, never from in-process handles;
- ``verifier``: an action-specific verification hook run by the executor
  after the built-in scientific/map checks (fail-closed);
- ``input_refs`` / ``output_refs``: typed-ref vocabulary names (see
  :data:`paleo_workbench.providers.contracts.TYPED_REFS`) the action
  consumes/produces — documentation *and* a registry-level check;
- ``domain_tags``: free-form, lowercase retrieval tags.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

_ACTION_ID_RE = re.compile(r"^[a-z][a-z0-9_]*\.[a-z][a-z0-9_]*$")
_VERSION_RE = re.compile(r"^[0-9]+(\.[0-9]+){0,3}$")
_TAG_RE = re.compile(r"^[a-z0-9][a-z0-9_.-]*$")


class ActionRisk(str, Enum):
    READ = "read"          # observe workspace/catalog/context — no mutation
    COMPUTE = "compute"    # derive new data through providers/services
    WRITE = "write"        # mutate documents/versions via domain services
    DESTRUCTIVE = "destructive"  # purge/overwrite; never a default agent action


class ActionStatus(str, Enum):
    """Canonical terminal states of one action execution (Harness 2.0).

    Six states, no aliases: ``rejected`` means a *guard* refused the
    request before/at admission (schema, permission, context, resources);
    ``failed`` means the execution or its verification did not hold;
    ``unavailable`` means a required production capability is missing —
    never a fake result. ``degraded`` is a success carrying warnings.
    """

    SUCCESS = "success"
    DEGRADED = "degraded"
    FAILED = "failed"
    CANCELLED = "cancelled"
    REJECTED = "rejected"
    UNAVAILABLE = "unavailable"


#: Statuses an agent may treat as "the work happened" (data was produced).
#: Everything else is a terminal non-result.
POSITIVE_STATUSES = frozenset({ActionStatus.SUCCESS, ActionStatus.DEGRADED})

#: Default permission set for programmatic/headless contexts: READ + COMPUTE.
#: WRITE is granted explicitly (the app context grants it — a user session
#: sits behind the UI and writes go through domain services either way);
#: DESTRUCTIVE is not installable at all.
DEFAULT_PERMISSIONS = frozenset({ActionRisk.READ, ActionRisk.COMPUTE})


@dataclass(frozen=True, slots=True)
class ActionSpec:
    """Declarative contract of one harness action."""

    action_id: str
    description: str
    handler: Any = field(compare=False, default=None, repr=False)  # callable(ActionContext, dict) -> Any
    input_schema: dict[str, Any] = field(default_factory=dict)
    output_schema: dict[str, Any] = field(default_factory=dict)
    risk: ActionRisk = ActionRisk.READ
    category: str = "background.compute"  # TaskCategory value for admission
    resource_profile: dict[str, Any] = field(
        default_factory=lambda: {
            "estimated_cpu_cores": 0.5,
            "estimated_ram_bytes": 0,
            "estimated_vram_bytes": 0,
            "io_weight": 0.5,
        }
    )
    required_context: tuple[str, ...] = ()  # ActionContext attrs that must be present
    supports_cancel: bool = False
    provider_id: str | None = None  # when execution delegates to a capability provider
    side_effect_notes: str = ""
    # --- Harness 2.0 -------------------------------------------------------
    version: str = "1.0"  # action contract version; cache/receipt identity
    deterministic: bool = False  # same inputs+params+version → same outputs
    cacheable: bool = False  # may reuse a prior catalog-complete execution
    idempotent: bool = False  # re-execution with same inputs is safe
    verifier: Any = field(compare=False, default=None, repr=False)  # callable(payload, parameters, context) -> report
    domain_tags: tuple[str, ...] = ()
    input_refs: tuple[str, ...] = ()   # typed-ref vocabulary names consumed
    output_refs: tuple[str, ...] = ()  # typed-ref vocabulary names produced

    @property
    def domain(self) -> str:
        return self.action_id.split(".", 1)[0]

    def to_dict(self) -> dict[str, Any]:
        return {
            "action_id": self.action_id,
            "description": self.description,
            "input_schema": self.input_schema,
            "output_schema": self.output_schema,
            "risk": self.risk.value,
            "category": self.category,
            "resource_profile": dict(self.resource_profile),
            "required_context": list(self.required_context),
            "supports_cancel": self.supports_cancel,
            "provider_id": self.provider_id,
            "side_effect_notes": self.side_effect_notes,
            "version": self.version,
            "deterministic": self.deterministic,
            "cacheable": self.cacheable,
            "idempotent": self.idempotent,
            "domain_tags": list(self.domain_tags),
            "input_refs": list(self.input_refs),
            "output_refs": list(self.output_refs),
        }

    def tool_schema(self) -> dict[str, Any]:
        """Agent tool definition (OpenAI/Gemini function-calling shape),
        derived — the schema is never re-authored anywhere else."""
        return {
            "type": "function",
            "function": {
                "name": self.action_id.replace(".", "__"),
                "description": (
                    f"{self.description} [risk: {self.risk.value}"
                    + (f"; via provider {self.provider_id}" if self.provider_id else "")
                    + "]"
                ),
                "parameters": self.input_schema
                or {"type": "object", "properties": {}, "additionalProperties": False},
            },
        }


def validate_schema_shape(schema: Any, *, _path: str = "schema") -> list[str]:
    """Recursively check that ``schema`` is a well-formed JSON-schema subset.

    Registration-time gate: unknown ``type`` names, non-dict ``properties``,
    non-list ``required``, non-dict ``items`` and misplaced bounds are
    reported here instead of silently passing at execution time.
    """
    problems: list[str] = []
    if not isinstance(schema, dict):
        return [f"{_path}: must be a dict (JSON schema)"]
    expected = schema.get("type")
    known = {"object", "array", "string", "integer", "number", "boolean", "null"}
    if expected is not None:
        names = [str(t) for t in expected] if isinstance(expected, list) else [str(expected)]
        for name in names:
            if name not in known:
                problems.append(f"{_path}: unknown type {name!r}")
    if "enum" in schema and not isinstance(schema["enum"], list):
        problems.append(f"{_path}: enum must be a list")
    for bound in ("minimum", "maximum"):
        if bound in schema and not isinstance(schema[bound], (int, float)):
            problems.append(f"{_path}: {bound} must be numeric")
    for bound in ("minItems", "maxItems"):
        if bound in schema and not isinstance(schema[bound], int):
            problems.append(f"{_path}: {bound} must be an integer")
    properties = schema.get("properties")
    if properties is not None:
        if not isinstance(properties, dict):
            problems.append(f"{_path}.properties: must be a dict")
        else:
            for key, sub in properties.items():
                problems.extend(validate_schema_shape(sub, _path=f"{_path}.properties.{key}"))
    required = schema.get("required")
    if required is not None and not (isinstance(required, list) and all(isinstance(r, str) for r in required)):
        problems.append(f"{_path}.required: must be a list of strings")
    items = schema.get("items")
    if items is not None:
        if isinstance(items, dict):
            problems.extend(validate_schema_shape(items, _path=f"{_path}.items"))
        else:
            problems.append(f"{_path}.items: must be a dict")
    additional = schema.get("additionalProperties")
    if additional is not None and not isinstance(additional, bool):
        problems.append(f"{_path}.additionalProperties: must be a boolean")
    return problems


def validate_action_spec(spec: ActionSpec) -> list[str]:
    problems: list[str] = []
    if not _ACTION_ID_RE.match(spec.action_id or ""):
        problems.append(
            f"action_id {spec.action_id!r} must be '<domain>.<name>' lowercase dotted"
        )
    if not (spec.description or "").strip():
        problems.append("description must be non-empty")
    if not isinstance(spec.risk, ActionRisk):
        problems.append(f"risk {spec.risk!r} is not an ActionRisk")
    if spec.handler is None and spec.provider_id is None:
        problems.append("spec needs a handler or a provider_id to be executable")
    schema = spec.input_schema
    if not isinstance(schema, dict):
        problems.append("input_schema must be a dict (JSON schema)")
    elif schema.get("type") not in (None, "object"):
        problems.append("input_schema must describe an object at the top level")
    if isinstance(schema, dict):
        problems.extend(validate_schema_shape(schema, _path="input_schema"))
    if spec.output_schema:
        problems.extend(validate_schema_shape(spec.output_schema, _path="output_schema"))
    if spec.version and not _VERSION_RE.match(spec.version):
        problems.append(f"version {spec.version!r} must be numeric dotted (e.g. 1.0)")
    if spec.cacheable and not spec.deterministic:
        problems.append("cacheable requires deterministic (a cache hit replays the execution)")
    if spec.cacheable and not spec.output_refs:
        problems.append(
            "cacheable requires declared output_refs: a cache hit must "
            "re-materialise from catalog-resolvable outputs"
        )
    if spec.verifier is not None and not callable(spec.verifier):
        problems.append("verifier must be callable(payload, parameters, context)")
    for tag in spec.domain_tags:
        if not isinstance(tag, str) or not _TAG_RE.match(tag):
            problems.append(f"domain_tag {tag!r} must match {_TAG_RE.pattern}")
    from paleo_workbench.providers.contracts import TYPED_REFS

    for role, refs in (("input_refs", spec.input_refs), ("output_refs", spec.output_refs)):
        for ref in refs:
            if ref not in TYPED_REFS:
                problems.append(f"{role} entry {ref!r} is not a known typed ref {sorted(TYPED_REFS)}")
    profile = spec.resource_profile
    if not isinstance(profile, dict):
        problems.append("resource_profile must be a dict")
    else:
        if float(profile.get("estimated_cpu_cores", 0.5)) <= 0:
            problems.append("resource_profile.estimated_cpu_cores must be > 0")
        if float(profile.get("io_weight", 0.5)) < 0:
            problems.append("resource_profile.io_weight must be >= 0")
        for key in ("estimated_temp_bytes",):
            if key in profile and not isinstance(profile[key], (int, float)):
                problems.append(f"resource_profile.{key} must be numeric")
    return problems
